### PR TITLE
[rom-e2e] Complete the SFDP verification

### PR DESCRIPTION
**Pull request by cfrantz**
1. Improve the SFDP parser to understand the the Basic Flash Parameter Table up to revision F.
2. Update the rom e2e test to check the relevant base and rev_b fields.
3. Correct an error in the ROM's SFDP: deep powerdown is not supported.

### CODE REVIEW

**Comment by cfrantz** [comment]
Fixes #14457.

File: sw/host/opentitanlib/src/spiflash/sfdp.rs [hunk 1]
```diff
@@ -105,82 +130,206 @@ impl TryFrom<&[u8]> for InternalJedecParams {
...
+
+    // JESD216B pp. 23-24.
+    field!(prohibited_pgm_suspend -> u8, 12, 0, 4);
+    field!(prohibited_erase_suspend -> u8, 12, 4, 4);
+    field!(pgm_resume_to_suspend -> u32, 12, 9, 4);
+    field!(suspend_pgm_latency_time -> u32, 12, 13, 5);
+    field!(suspend_pgm_latency_unit -> usize, 12, 18, 2);
+    field!(erase_resume_to_suspend -> u32, 12, 20, 4);
+    field!(suspend_erase_latency_time -> u32, 12, 24, 5);
+    field!(suspend_erase_latency_unit -> usize, 12, 19, 2);
```
Comment (alphan) [inline sw/host/opentitanlib/src/spiflash/sfdp.rs:245]: typo: 19 -> 29

File: sw/device/silicon_creator/lib/drivers/spi_device.c [hunk 2]
```diff
@@ -285,12 +285,14 @@ static_assert(kBfptTablePointer % sizeof(uint32_t) == 0,
 /**
  * BFPT 12th Word
  * --------------
- * [31: 9]: Erase/program suspend/resume (not supported, 0x0)
+ * [31:31]: Deep powerdown support (not supported, 0x1)
```
Comment (alphan) [inline sw/device/silicon_creator/lib/drivers/spi_device.c:288]: Thanks for catching this! But I think deep powerdown is the MSb of the 14th DWORD, this should be "suspend/resume support"

**Review by alphan — APPROVED** [review]
Thanks @cfrantz!

File: sw/host/opentitanlib/src/spiflash/sfdp.rs [hunk 3]
```diff
@@ -105,82 +130,206 @@ impl TryFrom<&[u8]> for InternalJedecParams {
...
+    field!(write_en_opcode -> bool, 1, 4, 1);
+    field!(erase_opcode_4kib -> u8, 1, 8, 8);
+    field!(support_fast_read_112 -> bool, 1, 16, 1);
+    field!(address_modes -> SupportedAddressModes, 1, 17, 2);
+    field!(support_double_rate_clocking -> bool, 1, 19, 1);
+    field!(support_fast_read_122 -> bool, 1, 20, 1);
+    field!(support_fast_read_144 -> bool, 1, 21, 1);
+    field!(support_fast_read_114 -> bool, 1, 22, 1);
+
+    field!(density -> u32, 2, 0, 32);
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/spiflash/sfdp.rs:178]: Later code accounts for this issue, but it's clearer if its represented as a bool field within InternalJedecParams.  Fixed.

File: sw/host/opentitanlib/src/spiflash/sfdp.rs [hunk 4]
```diff
@@ -105,82 +130,206 @@ impl TryFrom<&[u8]> for InternalJedecParams {
...
+    field!(wait_states_112 -> u8, 4, 16, 5);
+    field!(mode_bits_112 -> u8, 4, 21, 3);
+    field!(read_opcode_112 -> u8, 4, 22, 8);
+
+    field!(support_fast_read_222 -> bool, 5, 0, 1);
+    field!(support_fast_read_444 -> bool, 5, 4, 1);
+
+    field!(wait_states_222 -> u8, 6, 16, 5);
+    field!(mode_bits_222 -> u8, 6, 21, 3);
+    field!(read_opcode_222 -> u8, 6, 22, 8);
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/spiflash/sfdp.rs:199]: Yep.  A long standing bug in this code.  Same for `read_opcode_444`.  Fixed.

File: sw/host/opentitanlib/src/spiflash/sfdp.rs (hunk 1 above)
Comment (cfrantz) [reply sw/host/opentitanlib/src/spiflash/sfdp.rs:245]: Fixed.

File: sw/host/opentitanlib/src/spiflash/sfdp.rs [hunk 5]
```diff
@@ -105,82 +130,206 @@ impl TryFrom<&[u8]> for InternalJedecParams {
...
+    field!(status_busy_polling -> u8, 14, 2, 6);
+    field!(exit_deep_powerdown_time -> u32, 14, 8, 5);
+    field!(exit_deep_powerdown_unit -> usize, 14, 13, 2);
+    field!(exit_deep_powerdown_instruction -> u8, 14, 15, 8);
+    field!(enter_deep_powerdown_instruction -> u8, 14, 23, 8);
+    field!(deep_powerdown_supported -> bool, 14, 31, 1);
+
+    // JESD216B pp. 26-27.
+    field!(mode_444_disable -> u8, 15, 0, 4);
+    field!(mode_444_enable -> u8, 15, 8, 4);
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/opentitanlib/src/spiflash/sfdp.rs:264]: Good catch. Fixed.

File: sw/host/tests/rom/e2e_bootstrap_entry/src/main.rs [hunk 6]
```diff
@@ -251,6 +252,57 @@ fn test_sfdp(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
...
+    assert_eq!(rev_b.page_size, 256);
+    assert_eq!(rev_b.page_program_time.typical, parse_duration("768 us")?);
+    assert_eq!(rev_b.byte_program_time.typical, parse_duration("48 us")?);
+    assert_eq!(
+        rev_b.additional_byte_program_time.typical,
+        parse_duration("48 us")?
+    );
+    assert_eq!(rev_b.chip_erase_time.typical, parse_duration("192 ms")?);
+    // FIXME(cfrantz): Is this a bug in the ROM's SFDP table?
+    //assert_eq!(rev_b.suspend_resume_supported, false);
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply sw/host/tests/rom/e2e_bootstrap_entry/src/main.rs]: If I re-order the commits so that the fix to the ROM is first, then yes. :)

**Review by alphan — APPROVED** [review]
Thanks @cfrantz!

**Pull request merged**